services/poll_test.go

```diff
@@ -0,0 +1,128 @@
+package services
+
+import (
+	"context"
+	"errors"
+	"reflect"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ceramicnetwork/go-cas/models"
+)
+
+var originalCheckpoint = time.Now().Add(-time.Hour)
+
+type SpyAnchorRepository struct {
+	models.AnchorRepository
+	receivedCheckpoints []time.Time
+}
+
+func (f *SpyAnchorRepository) GetRequests(status models.RequestStatus, since time.Time, limit int) ([]*models.AnchorRequestMessage, error) {
+	f.receivedCheckpoints = append(f.receivedCheckpoints, since)
+
+	return []*models.AnchorRequestMessage{
+		{CreatedAt: since.Add(time.Minute * 1)},
+		{CreatedAt: since.Add(time.Minute * 2)},
+	}, nil
+}
+
+type StubStateRepository struct {
+	models.StateRepository
+	checkpoint time.Time
+}
+
+func (f *StubStateRepository) GetCheckpoint(CheckpointType models.CheckpointType) (time.Time, error) {
+	return f.checkpoint, nil
+}
+
+func (f *StubStateRepository) UpdateCheckpoint(checkpointType models.CheckpointType, time time.Time) (bool, error) {
+	f.checkpoint = time
+	return true, nil
+}
+
+type FakeReadyPublisher struct {
+	messages    chan any
+	numAttempts int
+	errorOn     int
+}
+
+func (f *FakeReadyPublisher) SendMessage(ctx context.Context, event any) (string, error) {
+	select {
+	case <-ctx.Done():
+		return "", errors.New("context cancelled")
+	default:
+		f.numAttempts = f.numAttempts + 1
+		if f.numAttempts == f.errorOn {
+			return "", errors.New("TestError")
+		}
+		f.messages <- event
+		return "msgId", nil
+	}
+
+}
+
+func waitForMesssages(messageChannel chan any, n int) []any {
+	messages := make([]any, n)
+	for i := 0; i < n; i++ {
+		message := <-messageChannel
+		messages[i] = message
+	}
+	return messages
+}
+
+// if no requests do not publish anything and try again after a second
+
+func TestPoller(t *testing.T) {
+	tests := map[string]struct {
+		readyPublisher            *FakeReadyPublisher
+		expectedCheckpoints       []time.Time
+		expectedCurrentCheckpoint time.Time
+	}{
+		"Can poll": {
```

Comment:
Would you mind terribly adding some comments for the timeline each test is checking for 😅?

My ADHD brain just can't follow, though I see what you're testing for.

e.g.
- T0s: Test starts
- T60s: First message sent
and so on

Reply:
done! Channels can quite hard to follow (the libp2p library makes more sense now)

Reply:
Ty, looks good!